File: meshPruning.mjs
// @ts-check
import util from 'util';
import { Heap, HeapValue } from './cHeap/Heap.mjs';

const COST = {
    NOT_FOUND:5000,
    SWITCH:80,
    MESSAGE:2
};
const INIT_WEIGHT = 1.5;
const CUTOFF_DISTANCE = 10;
 
const LEARNING_RATE = 0.003;

class node{
    heapValue = undefined;
    switchEnabled = true;
    lastSwitchEnabled = true;
    /**
     * @type {Array<{
     *      node:node,
     *      cost:number
     * }>}
     */
    neighbors = [];
    minCost = Infinity;
    weight = INIT_WEIGHT;
    visited = false;
    /**
     * 
     * @param {{
     *      coordination:{
     *          x:number,
     *          y:number
     *      },
     *      id:number
     * }} params 
     */
    constructor(params) {
        this.coordination = params.coordination;
        this.id = params.id;
    }

    reset(){
        this.heapValue = undefined;
        this.minCost = Infinity;
        this.visited = false;
    }

    setSwitch(){
        this.lastSwitchEnabled = this.switchEnabled;
        this.switchEnabled = this.sigmoid(this.weight) >= Math.random();
    }

    /**
     * @param {Heap} minHeap
     */
    visitNeighbors(minHeap){
        if(this.switchEnabled){
            this.neighbors.forEach(n=>{
                if(!n.node.visited){
                    let newCost = this.minCost + n.cost;
                    if(newCost < n.node.minCost){
                        n.node.minCost = newCost;
                        if(n.node.heapValue === undefined){
                            n.node.heapValue = new HeapValue(n.node);
                            minHeap.add(n.node.heapValue);
                        }else{
                            minHeap.adjust(n.node.heapValue);
                        }
                    }
                }
            });
        }
        this.visited = true;
    }

    /**
     * @param {number} delta <0 if cost is reduced 
     * if delta < 0 && switchEnabled, increase weight
     */
    updateWeight(delta){
        if(this.switchEnabled !== this.lastSwitchEnabled){
            this.weight += delta * (this.switchEnabled?-1:1);
        }
    }

    /**
     * @param {number} v 
     */
    sigmoid(v){
        return 1/(1+Math.exp(-v));
    }

    dump(){
        return {
            id:this.id,
            minCost:this.minCost,
            switchEnabled:this.switchEnabled,
        };
    }
}

class graph{
    /**
     * @type {Array<node>}
     */
    nodes = [];

    minHeap = new Heap(/** @param {node} A @param {node} B */(A,B)=>A.minCost>B.minCost);

    /**
     * 
     * @param {{
     *      nNode:number,
     *      fieldX:number,
     *      fieldY:number
     * }} params 
     */
    constructor(params){
        this.nodes.push(new node({
            id:0,
            coordination:{x:0,y:0}
        }));
        for(let i = 1;i<params.nNode;i++){
            this.nodes.push(new node({
                id:i,
                coordination:{
                    x:Math.random()*params.fieldX-params.fieldX/2,
                    y:Math.random()*params.fieldY-params.fieldY/2
                }
            }));
        }
        /** init neighbors */
        this.nodes.forEach(A=>{
            this.nodes.forEach(B=>{
                if(A.id !== B.id){
                    let d = this.getNodeDistance(A,B);
                    if(d < CUTOFF_DISTANCE){
                        let cost = this.distanceToCost(d);
                        if(!(A.id===0 || B.id===0)){
                            cost += COST.SWITCH;
                        }
                        A.neighbors.push({
                            node:B,
                            cost:cost
                        });
                    }
                }
            });
        });
    };

    /**
     * @param {node} A 
     * @param {node} B 
     */
    getNodeDistance(A,B) {
        return Math.sqrt((A.coordination.x-B.coordination.x)**2 + (A.coordination.y-B.coordination.y)**2);
    };

    /**
     * @param {number} d
     */
    distanceToCost(d){
        return (d*5+30) >>> 0;
    };

    /**
     * @returns {node | undefined}
     */
    getNextNodeToVisit(){
        let v = this.minHeap.pop()?.value;
        return v;
    }

    reset(){
        this.nodes.forEach(n=>{n.reset();});
        /** always set root cost to 0 */
        this.nodes[0].minCost = 0;
    }

    /**
     * @param {boolean} [forceEnable] 
     */
    setSwitch(forceEnable){
        if(forceEnable === true){
            this.nodes.forEach(n=>{n.switchEnabled = true});
        }else{
            this.nodes.forEach(n=>{n.setSwitch();});
        }
        /** always enable switch for the root */
        this.nodes[0].switchEnabled = true;
    }

    calculateCost(){
        this.minHeap = new Heap(/** @param {node} A @param {node} B */(A,B)=>A.minCost>B.minCost);
        let n = this.nodes[0];
        do {
            n.visitNeighbors(this.minHeap);
        } while ((n = this.getNextNodeToVisit())!==undefined);
        let pathCost = 0;
        this.nodes.forEach(n=>{
            pathCost += n.minCost===Infinity?COST.NOT_FOUND:n.minCost;
        });
        pathCost /= this.nodes.length;
        let nSwitch = 0;
        this.nodes.forEach(n=>{
            nSwitch += n.switchEnabled?1:0;
        });
        let msgCost = nSwitch*COST.MESSAGE;
        return pathCost + msgCost;
    }

    /**
     * @param {number} delta 
     */
    updateWeight(delta){
        this.nodes.forEach(n=>{n.updateWeight(delta);});
    }

    dump(){
        return this.nodes.map(n=>(n.dump()))
    };
}

/** @type {number} */
let lastCost = undefined;
let g = new graph({nNode:1000,fieldX:30,fieldY:40});
g.reset();
g.setSwitch(true);
lastCost = g.calculateCost();
console.log(`Inital cost: ${lastCost}`);

console.log(util.inspect(g.dump(),{maxArrayLength:null,depth:null,showHidden:false}));

for(let i=0;i<20000;i++){
    g.reset();
    g.setSwitch();
    let newCost = g.calculateCost();
    console.log(`Round ${i} cost: ${newCost}`);
    let delta = (newCost-lastCost)*LEARNING_RATE;
    lastCost = newCost;
    g.updateWeight(delta);
}


console.log(util.inspect(g.dump(),{maxArrayLength:null,depth:null,showHidden:false}));

File: mesh_pruning.c
#include "cHeap/heap.h"
#include <stdbool.h>
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define COST_NOT_FOUND (5000)
#define COST_SWITCH (80)
#define COST_MESSAGE (2)
#define INIT_WEIGHT (1.5f)
#define CUTOFF_DISTANCE (10)
#define LEARNING_RATE (0.003f)

float sigmoid(float v)
{
    return 1.0f/(1.0f + expf(-v));
}

float randf()
{
    return (((float)rand())/((float)RAND_MAX));
}

typedef struct mesh_node_neighbor_s mesh_node_neighbor_t;

typedef struct mesh_node_s
{
    heap_value_handle_t heap_value;
    bool switch_enabled;
    bool last_switch_enabled;
    mesh_node_neighbor_t* neighbors;
    int num_neighbors;
    float min_cost;
    float weight;
    bool visited;
    float x;
    float y;
    int id;
} mesh_node_t;

struct mesh_node_neighbor_s
{
    mesh_node_t* node;
    float cost;
};

/* mesh node will be static, thus only init them here */
void mesh_node_init(mesh_node_t* node,int id,float x,float y)
{
    memset(node,0,sizeof(mesh_node_t));
    node->heap_value = NULL;
    node->switch_enabled = true;
    node->last_switch_enabled = true;
    node->neighbors = NULL;
    node->min_cost = INFINITY;
    node->weight = INIT_WEIGHT;
    node->visited = false;
    node->id = id;
    node->x = x;
    node->y = y;
}

/* free neighbors */
void mesh_node_deinit(mesh_node_t* node)
{
    if(node->neighbors != NULL)
    {
        free(node->neighbors);
        node->neighbors = NULL;
        node->num_neighbors = 0;
    }
}

void mesh_node_reset(mesh_node_t* node)
{
    /* the heap should be freed as a whole */
    node->heap_value = NULL;
    node->min_cost = INFINITY;
    node->visited = false;
}

void mesh_node_set_switch(mesh_node_t* node)
{
    node->last_switch_enabled = node->switch_enabled;
    node->switch_enabled = sigmoid(node->weight) >= randf();
}

void mesh_node_visit_neighbors(mesh_node_t* node,heap_handle_t min_heap)
{
    if(node->switch_enabled){
        for(int i=0;i<node->num_neighbors;i++)
        {
            mesh_node_neighbor_t* neighbor = &(node->neighbors)[i];
            if(!neighbor->node->visited)
            {
                float new_cost = node->min_cost + neighbor->cost;
                if(new_cost < neighbor->node->min_cost)
                {
                    neighbor->node->min_cost = new_cost;
                    if(neighbor->node->heap_value == NULL)
                        neighbor->node->heap_value = heap_add(min_heap,neighbor->node);
                    else
                        heap_update(min_heap,neighbor->node->heap_value);
                }
            }
        }
    }
    node->visited = true;
}

void mesh_node_update_weight(mesh_node_t* node,float delta)
{
    if(node->switch_enabled != node->last_switch_enabled)
        node->weight += delta * (node->switch_enabled?-1.0f:1.0f);
}

void mesh_node_dump(mesh_node_t* node)
{
    printf("id:%d,\tmin cost:%.0f,\tswitch enabled:%s\n",
            node->id,node->min_cost,node->switch_enabled?"Y":"N");
}

float mesh_node_get_dis(mesh_node_t* A,mesh_node_t* B)
{
    return sqrtf(powf(A->x-B->x,2)+powf(A->y-B->y,2));
}

float mesh_node_dis_to_cost(float d)
{
    return roundf(d*5.0f+30.0f);
}

typedef struct
{
    mesh_node_t* nodes;
    int n_node;
    heap_handle_t min_heap;
} mesh_graph_t;

mesh_graph_t* mesh_graph_new(int n_node,float x,float y)
{
    mesh_graph_t* graph = malloc(sizeof(mesh_graph_t));
    if(!graph)
        goto fail;
    /* min heap is inited on use, not here. */
    graph->min_heap = NULL;
    graph->n_node = n_node;
    graph->nodes = malloc(sizeof(mesh_node_t)*n_node);
    if(!graph->nodes)
        goto fail;
    srand(time(NULL));
    int i = 0;
    mesh_node_init(&graph->nodes[i],i,0,0);
    i++;
    for(;i<n_node;i++)
    {
        mesh_node_init(&graph->nodes[i],i,(randf()-0.5f)*x,(randf()-0.5f)*y);
    }
    for(i=0;i<n_node;i++)
    {
        mesh_node_t* A = &graph->nodes[i];
        for(int j=0;j<n_node;j++)
        {
            mesh_node_t* B = &graph->nodes[j];
            if(i != j)
            {
                float d = mesh_node_get_dis(A,B);
                if(d < CUTOFF_DISTANCE)
                {
                    float cost = mesh_node_dis_to_cost(d);
                    if(!(i==0||j==0))
                        cost += COST_SWITCH;
                    A->num_neighbors ++;
                    A->neighbors = realloc(A->neighbors,sizeof(mesh_node_neighbor_t)*A->num_neighbors);
                    if(!A->neighbors)
                        exit(1);
                    mesh_node_neighbor_t* neighbor = &(A->neighbors)[A->num_neighbors-1];
                    neighbor->cost = cost;
                    neighbor->node = B;
                }
            }
        }
    }
    return graph;
fail:
    exit(1);
    return NULL;
}

void mesh_graph_free(mesh_graph_t* graph)
{
    if(!graph)
        return;
    if(graph->min_heap)
    {
        heap_free(graph->min_heap,NULL,NULL);
        graph->min_heap = NULL;
    }
    if(graph->nodes)
    {
        for(int i=0;i<graph->n_node;i++)
            mesh_node_deinit(&graph->nodes[i]);
        free(graph->nodes);
        graph->nodes = NULL;
    }
    free(graph);
}

void mesh_graph_reset(mesh_graph_t* graph)
{
    for(int i=0;i<graph->n_node;i++)
        mesh_node_reset(&graph->nodes[i]);
    graph->nodes[0].min_cost = 0;
}

void mesh_graph_set_switch(mesh_graph_t* graph, bool force_enable)
{
    if(force_enable)
    {
        for(int i=0;i<graph->n_node;i++)
            graph->nodes[i].switch_enabled = true;
    }
    else
    {
        for(int i=0;i<graph->n_node;i++)
            mesh_node_set_switch(&graph->nodes[i]);
    }
    graph->nodes[0].switch_enabled = true;
}

bool mesh_node_cost_compare(void* A,void* B)
{
    mesh_node_t* node_A = A;
    mesh_node_t* node_B = B;
    return node_A->min_cost > node_B->min_cost;
}

float mesh_graph_get_cost(mesh_graph_t* graph)
{
    float total_cost = 0.0f;
    if(graph->min_heap)
        heap_free(graph->min_heap,NULL,NULL);
    graph->min_heap = heap_create(mesh_node_cost_compare);
    if(!graph->min_heap)
        goto fail;
    
    mesh_node_t* n = &graph->nodes[0];
    for(;;)
    {
        mesh_node_visit_neighbors(n,graph->min_heap);
        n = heap_pop(graph->min_heap);
        if(!n)
            break;
    }
    float path_cost = 0.0f;
    for(int i=0;i<graph->n_node;i++)
    {
        float node_cost = graph->nodes[i].min_cost;
        if(isinff(node_cost))
            node_cost = COST_NOT_FOUND;
        path_cost += node_cost;
    }
        
    path_cost /= graph->n_node;
    int n_switch = 0;
    for(int i=0;i<graph->n_node;i++)
        n_switch += graph->nodes[i].switch_enabled;
    float msg_cost = ((float)n_switch) * COST_MESSAGE;
    total_cost = path_cost + msg_cost;

    return total_cost;
fail:
    exit(1);
    return NAN;
}

void mesh_graph_update_weight(mesh_graph_t* graph, float delta)
{
    for(int i=0;i<graph->n_node;i++)
        mesh_node_update_weight(&graph->nodes[i],delta);
}

void mesh_graph_dump(mesh_graph_t* graph)
{
    for(int i=0;i<graph->n_node;i++)
        mesh_node_dump(&graph->nodes[i]);
}

int main(int argc, char const *argv[])
{
    mesh_graph_t* g = mesh_graph_new(1000,30,40);
    
    mesh_graph_reset(g);
    mesh_graph_set_switch(g,true);
    float last_cost = mesh_graph_get_cost(g);
    mesh_graph_dump(g);
    printf("Initial cost: %.2f\n",last_cost);

    for(int i=0;i<20000;i++)
    {
        mesh_graph_reset(g);
        mesh_graph_set_switch(g,false);
        float new_cost = mesh_graph_get_cost(g);
        printf("Round %d cost: %.2f\n",i,new_cost);
        float delta = (new_cost - last_cost)*LEARNING_RATE;
        last_cost = new_cost;
        mesh_graph_update_weight(g,delta);
    }

    mesh_graph_dump(g);

    mesh_graph_free(g);
    return 0;
}
